contracts/golos.worker/main.cpp

```diff
@@ -0,0 +1,1088 @@
+    eosio_assert(tspec != proposal_ptr->tspec_apps.end(), "technical specification doesn't exist");
+    require_app_member(tspec->author);
+    eosio_assert(tspec->votes.upvotes.empty(), "technical specification bid can't be deleted because it already has been upvoted"); //Technical Specification 1.e
+
+    get_proposals().modify(proposal_ptr, tspec->author, [&](auto &o) {
+      o.tspec_apps.erase(get_tspec(o, tspec_app_id));
+    });
+  }
+
+  /**
+   * @brief votetspec votes for the technical specification application
+   * @param proposal_id proposal ID
+   * @param tspec_app_id technical specification application
+   * @param author voting account name
+   * @param vote 1 - for the positive vote, 0 - for the negative vote. Look at a voting_module_t::vote_t
+   * @param comment_id comment ID of the comment that will be attached as a description to the vote
+   * @param comment attached comment data
+   */
+  /// @abi action
+  void votetspec(proposal_id_t proposal_id, tspec_id_t tspec_app_id, account_name author, uint8_t vote, comment_id_t comment_id, const comment_data_t &comment)
+  {
+    LOG("proposal_id: %, tpsec_id: %, author: %, vote: %", proposal_id, tspec_app_id, ACCOUNT_NAME_CSTR(author), (int)vote);
+
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_APP, "invalid state " __FILE__ ":" TOSTRING(__LINE__));
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+
+    const auto tspec_ptr = get_tspec(*proposal_ptr, tspec_app_id);
+    eosio_assert(tspec_ptr != proposal_ptr->tspec_apps.end(), "technical specification applicatoin doesn't exist");
+    require_app_delegate(author);
+    eosio_assert(voting_time_s + tspec_ptr->created.to_time_point().sec_since_epoch() >= now(), "voting time is over");
+
+    get_proposals().modify(proposal_ptr, author, [&](auto &o) {
+      auto tspec = get_tspec(o, tspec_app_id);
+      tspec->votes.vote(author, static_cast<voting_module_t::vote_value_t>(vote));
+
+      if (!comment.text.empty())
+      {
+        tspec->comments.add(comment_id, author, comment);
+      }
+
+      switch (vote)
+      {
+      case voting_module_t::VOTE_UP:
+        if (tspec->votes.upvotes.size() >= witness_count_51)
+        {  
+          //TODO: check that all voters are delegates in this moment
+          choose_proposal_tspec(o, *tspec, author);
+        }
+        break;
+      case voting_module_t::VOTE_DOWN:
+        break;
+      }
+    });
+  }
+
+  /**
+   * @brief publishtspec publish a final tehcnical specification
+   * @param proposal_id proposal ID
+   * @param data technical specification details
+   */
+  /// @abi action
+  void publishtspec(proposal_id_t proposal_id, const tspec_data_t &data)
+  {
+    LOG("proposal_id: %", proposal_id);
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_CREATE, "invalid proposal state");
```

Comment:
*TODO: now it allows push this action several times. can be used to edit. is it needed?*